### PR TITLE
Feature/producer consumer

### PR DESCRIPTION
**Pull request by myhndl**
Ready for review, but DO NOT MERGE!

Needs coordinated release with docs updates.

### CODE REVIEW

File: {{ cookiecutter.project_slug }}/docker/test-integration.yml [hunk 1]
```diff
@@ -1,8 +1,10 @@
-version: '3.4'
+version: '2'
```
Comment (mrtizmoatwork) [inline {{ cookiecutter.project_slug }}/docker/test-integration.yml]: why are you going with an older version of the dockerfile format?

File: {{ cookiecutter.project_slug }}/docker/test-integration.yml [hunk 2]
```diff
@@ -13,21 +15,82 @@ services:
...
+  test_redis_{{ cookiecutter.participant_name_slug }}:
+    image: "redis"
+    environment:
+      - ALLOW_EMPTY_PASSWORD=yes
+    ports:
+      - "6379:6379"
+
+  test_microengine-frontend_{{ cookiecutter.participant_name_slug }}:
+    image: "polyswarm/polyswarm-client:latest"
+    depends_on:
```
Comment (mrtizmoatwork) [inline {{ cookiecutter.project_slug }}/docker/test-integration.yml]: also depends on redis

File: {{ cookiecutter.project_slug }}/docker/test-integration.yml [hunk 3]
```diff
@@ -13,21 +15,82 @@ services:
...
       - "10"
+      - "--backend"
+      - "producer"
       - "--chains"
       - "side"
+
+  test_microengine-worker_{{ cookiecutter.participant_name_slug }}:
+    image: "{{ cookiecutter.project_slug }}"
+    depends_on:
+      - polyswarmd
```
Comment (mrtizmoatwork) [inline {{ cookiecutter.project_slug }}/docker/test-integration.yml]: also depends on redis

File: {{ cookiecutter.project_slug }}/docker/test-integration.yml (hunk 1 above)
Comment (myhndl) [reply {{ cookiecutter.project_slug }}/docker/test-integration.yml]: Base.yml is v2.

We could upgrade both, but that risks breaking e2e. Would need to test.

File: {{ cookiecutter.project_slug }}/docker/test-integration.yml (hunk 2 above)
Comment (myhndl) [reply {{ cookiecutter.project_slug }}/docker/test-integration.yml]: It has a dockerize wait on redis, which I think covers?

File: {{ cookiecutter.project_slug }}/docker/test-integration.yml (hunk 1 above)
Comment (mrtizmoatwork) [reply {{ cookiecutter.project_slug }}/docker/test-integration.yml]: should definitely put them all at least at version 3. But, we can do that stuff after we finish the e2e re-design, which will have solved all of the issues you are thinking about.

File: {{ cookiecutter.project_slug }}/docker/test-integration.yml (hunk 2 above)
Comment (mrtizmoatwork) [reply {{ cookiecutter.project_slug }}/docker/test-integration.yml]: well, if you put redis in the depends_on, then you probably do not need to dockerize wait stuff. But, we can change that later. @sbnetops is going to work on removing our usage of dockerize.

File: {{ cookiecutter.project_slug }}/docker/test-integration.yml (hunk 1 above)
Comment (myhndl) [reply {{ cookiecutter.project_slug }}/docker/test-integration.yml]: https://polyswarm.atlassian.net/browse/DN-2753

**Comment by myhndl** [comment]
@mrtizmoatwork `depends_on` is insufficient. You still need dockerize wait (or an alternative).

`depends_on` doesn't really do much for us, only assures that container A exists before B, not that the service in container A is ready.

**Comment by sbnetops** [comment]
`depends_on` is indeed not sufficient, our client should have some (re)connect/timeout feature to deal with this. I don't think we should rely on dockerize to do this, but instead have this built in within our packages.

**Pull request merged**